### PR TITLE
Iox #622 toml parser exceptions will crash roudi

### LINKED ISSUE #622: TOML parser exceptions will crash RouDi
## Required information

**Operating system:**
Ubuntu 18.04 LTS

**Compiler version:**
GCC 7.5.0

**Observed result or behaviour:**
RouDi can be crashed if malformed TOML Files are sent 

**Expected result or behaviour:**
RouDi is gracefully shutdown if TOML parser throws an exception

**Conditions where it occurred / Performed steps:**
Sent malformed TOML (attached) to RouDi and it'll crash.

**Details**
RouDi uses a 3rd party TOML parser to parse config files. This TOML parser throws exceptions whenever it cannot parse a file correctly. As a consequence, RouDi will crash/abort as it doesn't handle exceptions. Some examples of malformed TOML files that'll crash RouDi have been found during fuzz testing.

**Fix**
As discussed in the dev meeting on 2021-03-18. the used TOML parser is 3rd party and we won't change its code. However, we don't want RouDi to crash but shutdown gracefully with an error. The fix will be to wrap the parser call in a `try {} catch {}` block to detect exceptions and report an error.

**Related**
- Fuzz tool: [american fuzzy lop](https://lcamtuf.coredump.cx/afl/)
- https://github.com/eclipse-iceoryx/iceoryx/issues/558

**Attachements**
[crashes.zip](https://github.com/eclipse-iceoryx/iceoryx/files/6170559/crashes.zip)

### PR DESCRIPTION
**Pull request by marthtz**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/advanced/best-practice-for-testing.md

## Notes for Reviewer
Added `try() catch()` around TOML parser plus unit tests.
I'm getting the path to the input file via cmake as the path string is needed. Just have a look.
Btw, class `TomlRouDiConfigFileProvider` is not matching with its filename. You want me to fix it here?

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #622

### CODE REVIEW

File: iceoryx_posh/source/roudi/roudi_config_toml_file_provider.cpp [hunk 1]
```diff
@@ -66,7 +66,20 @@ iox::cxx::expected<iox::RouDiConfig_t, iox::roudi::RouDiConfigFileParseError> To
     }
     auto groupOfCurrentProcess = iox::posix::PosixGroup::getGroupOfCurrentProcess().getName();
 
-    std::shared_ptr<cpptoml::table> parsedFile = cpptoml::parse_file(m_customConfigFilePath.c_str());
+    std::shared_ptr<cpptoml::table> parsedFile{nullptr};
```
Comment (elfenpiff) [inline iceoryx_posh/source/roudi/roudi_config_toml_file_provider.cpp:69]: A `std::shared_ptr` is always initialized to `nullptr` so you do not need to do this explicitly. But for now, you can leave it as it is.
The initialization is only important for POD types like a pointer or int or something like this.

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_posh/source/roudi/roudi_config_toml_file_provider.cpp (hunk 1 above)
Comment (marthtz) [reply iceoryx_posh/source/roudi/roudi_config_toml_file_provider.cpp:69]: Thanks for the hint!

File: iceoryx_posh/include/iceoryx_posh/gateway/toml_gateway_config_parser.hpp [hunk 2]
```diff
@@ -35,11 +34,16 @@ enum TomlGatewayConfigParseError
...
     INCOMPLETE_CONFIGURATION,
     INCOMPLETE_SERVICE_DESCRIPTION,
-    INVALID_SERVICE_DESCRIPTION
+    INVALID_SERVICE_DESCRIPTION,
+    EXCEPTION_IN_PARSER
 };
 
-constexpr char TomlGatewayConfigParseErrorString[][64] = {
-    "FILE_NOT_FOUND", "INCOMPLETE_CONFIGURATION", "INCOMPLETE_SERVICE_DESCRIPTION", "INVALID_SERVICE_DESCRIPTION"};
+constexpr char TOML_GATEWAY_CONFIG_FILE_PARSE_ERROR_STRINGS[][64] = {"INVALID_STATE",
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/gateway/toml_gateway_config_parser.hpp:41]: this could be a `constexpr const char* TOML...STRINGS[] = {}`

**Review by elBoberido — APPROVED** [review]

File: iceoryx_posh/include/iceoryx_posh/gateway/toml_gateway_config_parser.hpp (hunk 2 above)
Comment (marthtz) [reply iceoryx_posh/include/iceoryx_posh/gateway/toml_gateway_config_parser.hpp:41]: Probably a general clean-up task as there're a lot of `constexpr char` in the code base!

**Pull request merged**